presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,397 @@
+                    case GREATER_THAN:
+                    case GREATER_THAN_OR_EQUAL:
+                        return new ComparisonExpression(GREATER_THAN, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
+                    default:
+                        throw new UnsupportedOperationException("Not yet implemented: " + operator);
+                }
+            }
+
+            if (literalVsRoundtripped < 0) {
+                // cast rounded up
+                switch (operator) {
+                    case EQUAL:
+                        return falseIfNotNull(cast.getExpression());
+                    case NOT_EQUAL:
+                        return trueIfNotNull(cast.getExpression());
+                    case IS_DISTINCT_FROM:
+                        return TRUE_LITERAL;
+                    case LESS_THAN:
+                    case LESS_THAN_OR_EQUAL:
+                        return new ComparisonExpression(LESS_THAN, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
+                    case GREATER_THAN:
+                    case GREATER_THAN_OR_EQUAL:
+                        if (sourceType.getRange().isPresent() && compare(sourceType, sourceType.getRange().get().getMax(), literalInSourceType) == 0) {
+                            return new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
+                        }
+                        return new ComparisonExpression(GREATER_THAN_OR_EQUAL, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
+                    default:
+                        throw new UnsupportedOperationException("Not yet implemented: " + operator);
+                }
+            }
+
+            return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
+        }
+
+        private boolean isInjectiveConversion(Type source, Type target)
+        {
+            if ((source.equals(BIGINT) && target.equals(DOUBLE)) ||
+                    (source.equals(BIGINT) && target.equals(REAL)) ||
+                    (source.equals(INTEGER) && target.equals(REAL))) {
+                // Not every BIGINT fits in DOUBLE/REAL due to 64 bit vs 53-bit/23-bit mantissa. Similarly,
+                // not every INTEGER fits in a REAL (32-bit vs 23-bit mantissa)
+                return false;
+            }
+
+            if (source instanceof DecimalType) {
+                int precision = ((DecimalType) source).getPrecision();
+
+                if (precision > 15 && target.equals(DOUBLE)) {
+                    // decimal(p,s) with p > 15 doesn't fit in a double without loss
+                    return false;
+                }
+
+                if (precision > 7 && target.equals(REAL)) {
```

Comment:
Should't this be `> 6` instead of 7?

https://en.wikipedia.org/wiki/Single-precision_floating-point_format says

> All integers with 6 or fewer significant decimal digits ... can be converted into an IEEE 754 floating-point value without loss of precision.

Reply:
That’s what I thought initially, but there was some issue with 7 digits. Maybe something in the decimal to real conversion implementation. Let me take another look

Reply:
I believe the comment in that page refers to *any* 6 significant digits. Single-precision format can actually represent 24 bits worth of significand, which is sufficient to store 7 digits for integer numbers:

```
presto> WITH t(x) AS (VALUES DECIMAL '9999999', DECIMAL '-9999999') SELECT typeof(x), CAST(x AS REAL) FROM t;
    _col0     |   _col1
--------------+------------
 decimal(7,0) |  9999999.0
 decimal(7,0) | -9999999.0
(2 rows)
```
I also tested every value between -9_999_999 and 9_999_999 to be sure and they all convert properly.

Reply:
Nice. Please add an explanatory comment here then.
Also, does the same apply to double precision? Ie can we raise `>15` threshold?